rust/agents/relayer/src/msg/serial_submitter.rs

```diff
@@ -54,329 +55,243 @@ use super::{
-        // still-unverified messages that have been in the verification queue
-        // for > threshold_time, move them back to the wait queue for further
-        // processing.
-
-        self.metrics
-            .run_queue_length_gauge
-            .set(self.run_queue.len() as i64);
-
-        // check if the next message is going to be processable
-        if let Some(Reverse(PendingMessage {
-            next_attempt_after: Some(retry_after),
-            ..
-        })) = self.run_queue.peek()
-        {
-            if Instant::now() < *retry_after {
-                return Ok(());
-            }
-        }
+#[instrument(skip_all)]
+async fn receive_task(
+    domain: HyperlaneDomain,
+    mut rx: mpsc::UnboundedReceiver<Box<DynPendingOperation>>,
+    prepare_queue: OpQueue,
+) -> Result<()> {
+    // Pull any messages sent to this submitter
+    while let Some(op) = rx.recv().await {
+        trace!(?op, "Received new operation");
+        // make sure things are getting wired up correctly; if this works in testing it
+        // should also be valid in production.
+        debug_assert_eq!(*op.domain(), domain);
+        prepare_queue.lock().await.push(Reverse(op));
+    }
+    bail!("Submitter receive channel was closed")
+}
 
-        // Pick the next message to try processing.
-        let mut msg = match self.run_queue.pop() {
-            Some(m) => m.0,
-            None => return Ok(()),
+#[instrument(skip_all)]
+async fn prepare_task(
+    tx_submit: mpsc::Sender<Box<DynPendingOperation>>,
+    prepare_queue: OpQueue,
+    metrics: SerialSubmitterMetrics,
+) -> Result<()> {
+    loop {
+        // Pick the next message to try preparing.
+        let next = {
+            let mut queue = prepare_queue.lock().await;
+            metrics.prepare_queue_length.set(queue.len() as i64);
+            queue.pop()
         };
-
-        match self.process_message(&msg).await {
-            Ok(true) => {
-                info!(msg=%msg.message, "Message processed");
-                self.record_message_process_success(&msg)?;
-                return Ok(());
+        let Some(Reverse(mut op)) = next else {
+            sleep(Duration::from_millis(200)).await;
+            continue;
+        };
+        trace!(?op, "Preparing operation");
+
+        match op.prepare().await {
+            PendingOperationResult::Success => {
+                debug!(?op, "Operation prepared");
+                metrics.ops_prepared.inc();
+                // this send will pause this task if the submitter is not ready to accept yet
+                tx_submit.send(op).await?;
+            }
+            PendingOperationResult::NotReady => {
+                // none of the operations are ready yet, so wait for a little bit
+                prepare_queue.lock().await.push(Reverse(op));
+                sleep(Duration::from_millis(200)).await;
+            }
+            PendingOperationResult::Reprepare => {
```

Comment:
nit: Wdyt of `Failed` as opposed to `Reprepare`?

Reply:
I prefer reprepare since it is a little more clear

Reply:
Works for me